spec/light-client/Blockchain.tla

```diff
@@ -0,0 +1,212 @@
+    \* the validators of the next block together with their voting powers,
+    \* i.e., a multi-set. We store the next validators instead of the hash.
+]
+
+(* A signed header is just a header together with a set of commits *)
+\* TODO: Commits is the set of PRECOMMIT messages
+SignedHeaders == BlockHeaders \X Commits
+
+VARIABLES
+    tooManyFaults,
+    (* whether there are more faults in the system than the blockchain can handle *)
+    height,
+    (* the height of the blockchain, starting with 0 *)
+    minTrustedHeight,
+    (* The global height of the oldest block that is younger than
+       the trusted period (AKA the almost rotten block).
+       In the implementation, this is the oldest block,
+       where block.bftTime + trustingPeriod >= globalClock.now. *)
+    blockchain,
+    (* A sequence of BlockHeaders,
+       which gives us the God's (or Daemon's) view of the blockchain. *)
+    Faulty
+    (* A set of faulty nodes, which can act as validators. We assume that the set
+       of faulty processes is non-decreasing. If a process has recovered, it should
+       connect using a different id. *)
+       
+(* all variables, to be used with UNCHANGED *)       
+vars == <<tooManyFaults, height, minTrustedHeight, blockchain, Faulty>>         
+
+(* The set of all correct nodes in a state *)
+Corr == AllNodes \ Faulty       
+
+(****************************** BLOCKCHAIN ************************************)
+(* in the future, we may extract it in a module on its own *)
+
+(* Compute the total voting power of a subset of validators Nodes,
+   whose individual voting powers are given by a function vp *)  
+RECURSIVE PowerOfSet(_, _)
+PowerOfSet(vp, Nodes) ==
+    IF Nodes = {}
+    THEN 0
+    ELSE LET node == CHOOSE n \in Nodes: TRUE IN
+        (* ASSERT(node \in DOMAIN vp) *)
+        vp[node] + PowerOfSet(vp, Nodes \ {node})
+
+TwoThirds(vp, set) ==
+    LET TP == PowerOfSet(vp, DOMAIN vp)
+        SP == PowerOfSet(vp, set)
+    IN
+    3 * SP > 2 * TP 
+
+IsCorrectPowerForSet(Flt, vp, set) ==
+    LET FV == Flt \intersect set
+        CV == set \ FV
+        CP == PowerOfSet(vp, CV)
+        FP == PowerOfSet(vp, FV)
+    IN
+    CP > 2 * FP \* 2/3 rule. Note: when FP = 0, this implies CP > 0.
+
+(* Is the voting power correct,
+   that is, more than 2/3 of the voting power belongs to the correct processes? *)
+IsCorrectPower(Flt, vp) ==
+    IsCorrectPowerForSet(Flt, vp, DOMAIN vp)
+    
+(* This is what we believe is the assumption about failures in Tendermint *)     
+FaultAssumption(Flt, mth, bc) ==
+    \A h \in mth..Len(bc):
+        IsCorrectPower(Flt, bc[h].NextVP)
+
+(* Append a new block on the blockchain.
+   Importantly, more than 2/3 of voting power in the next set of validators
+   belongs to the correct processes. *)       
+AppendBlock ==
+  LET last == blockchain[Len(blockchain)] IN
+  \E lastCommit \in (SUBSET (DOMAIN last.VP)) \ {{}},
+     NextV \in SUBSET AllNodes \ {{}}:
+     \E NextVP \in [NextV -> Powers]:
```

Comment:
This is hard to understand. It would help if we would have more comments.

Reply:
Fixed that.